internal/transports/p2p/network/address_book.go

```diff
@@ -0,0 +1,138 @@
+}
+
+// NewAddressBook creates and initializes a new AddressBook instance.
+func NewAddressBook(banDuration time.Duration, acceptLocalAddresses bool) *AddressBook {
+	// Set the address filter function based on whether local addresses are accepted
+	addrFilterFn := IsRoutable
+	if acceptLocalAddresses {
+		addrFilterFn = IsRoutableWithLocal
+	}
+
+	const addressesInitCapacity = 500
+	return &AddressBook{
+		addrs:        make([]*knownAddress, 0, addressesInitCapacity),
+		addrsLookup:  make(map[string]int, addressesInitCapacity),
+		banDuration:  banDuration,
+		addrFitlerFn: addrFilterFn,
+	}
+}
+
+// UpsertPeerAddr updates or adds a peer's address.
+func (a *AddressBook) UpsertPeerAddr(p *peer.Peer) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	pa := p.GetPeerAddr()
+	key, ka := a.findAddr(pa)
+
+	if ka != nil {
+		ka.peer = p
+	} else {
+		a.addAddr(key, &knownAddress{addr: pa, peer: p})
+	}
+}
+
+// UpsertAddrs updates or adds multiple addresses.
+func (a *AddressBook) UpsertAddrs(address []*wire.NetAddress) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	for _, addr := range address {
+		if !a.addrFitlerFn(addr) {
+			continue
+		}
+
+		key, ka := a.findAddr(addr)
+		// If the address is not found, add it to the AddressBook.
+		if ka == nil {
+			a.addAddr(key, &knownAddress{addr: addr})
+		} else if addr.Timestamp.After(ka.addr.Timestamp) {
+			// Otherwise, update the timestamp if the new one is newer.
+			ka.addr.Timestamp = addr.Timestamp
+		}
+	}
+}
+
+// BanAddr bans a network address. Ignores address if doesn't exist in the AddressBook.
+func (a *AddressBook) BanAddr(addr *wire.NetAddress) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	_, ka := a.findAddr(addr)
+	if ka != nil {
+		now := time.Now()
+		ka.banTimestamp = &now
+	}
+}
+
+// GetRandUnusedAddr returns a randomly chosen unused network address.
+func (a *AddressBook) GetRandUnusedAddr(tries uint) *wire.NetAddress {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	alen := len(a.addrs)
+	for i := uint(0); i < tries; i++ {
+		// #nosec G404
+		ka := a.addrs[rand.Intn(alen)]
+		if ka.peer == nil {
+			if ka.isBanned(a.banDuration) {
+				continue
+			}
+			return ka.addr
+		}
+	}
+
+	// return nil if no suitable address is found
+	return nil
+}
+
+func (a *AddressBook) findAddr(addr *wire.NetAddress) (string, *knownAddress) {
+	key := addrKey(addr)
+	addrIndex, ok := a.addrsLookup[key]
+
+	if ok {
+		return key, a.addrs[addrIndex]
+	}
+	return key, nil
+}
+
+func (a *AddressBook) addAddr(key string, addr *knownAddress) {
+	newItemIndex := len(a.addrs)
+
+	a.addrs = append(a.addrs, addr)
+	a.addrsLookup[key] = newItemIndex
+}
+
+func addrKey(addr *wire.NetAddress) string {
+	return fmt.Sprintf("%s:%d", addr.IP, addr.Port)
+}
+
+type knownAddress struct {
+	addr *wire.NetAddress
+	peer *peer.Peer
+
+	banTimestamp *time.Time
+}
+
+func (a *knownAddress) isBanned(duration time.Duration) bool {
+	return a.banTimestamp != nil && time.Since(*a.banTimestamp) < duration
```

Comment:
There's one more problem here. Once the peer is banned, we never unban him. Meaning, we never make `banTimestampt` to nil. So once the address is banned, even if the duration passes, we will still return that it's banned.

We probably need some handler for unbanning peers, either periodically every idk, 5-10 minutes, or during some operations, maybe even in this function as a part of checking if the address in banned.

Reply:
1. Firstly, it's a naive and straightforward ban implementation by design.
2. We don't need to "unban" an address because if the timestamp is older than the ban duration, the address is automatically no longer banned.

Reply:
Okay, you're right, I'm stupid and misunderstood the return statement

Reply:
All ban logic has been updated (along with the address book). Now, when we ban an address, it is moved to the banned collection, and a new goroutine is started. This goroutine is responsible for moving the address back to the free (unbanned, ready to use) collection when the ban is over